File: 363-max-sum-of-rectangle-no-larger-than-k/363-max-sum-of-rectangle-no-larger-than-k.py
from sortedcontainers import SortedSet
class Solution:
    def maxSumSubmatrix(self, matrix: List[List[int]], k: int) -> int:
        #
        nrow, ncol = len(matrix), len(matrix[0])
        def find_element(rowSum):
            sorted_sum = SortedSet([0])
            ans = -float('inf')
            running_sum = 0
            for num in rowSum:
                running_sum += num
                target = running_sum - k
                celing_index = sorted_sum.bisect_left(target)
                if celing_index < len(sorted_sum):
                    ans = max(ans, running_sum - sorted_sum[celing_index])
                sorted_sum.add(running_sum)
                if ans == k:
                    return ans
              #  print(ans, running_sum,sorted_sum[celing_index])
           # print(sorted_sum)
            return ans
        final_ans = -float('inf')
        for start_row in range(nrow):
            rowSum = [0] * ncol
            for cur_row in range(start_row, nrow):
                for col in range(ncol):
                    rowSum[col] += matrix[cur_row][col]
             #   print(rowSum)
                final_ans = max(final_ans, find_element(rowSum))
                if final_ans == k:
                    return final_ans
        return final_ans
                    

File: 363-max-sum-of-rectangle-no-larger-than-k/363-max-sum-of-rectangle-no-larger-than-k.java
class Solution {
    public int maxSumSubmatrix(int[][] matrix, int k) {
        int nrow = matrix.length, ncol = matrix[0].length;
        int res = Integer.MIN_VALUE;
        for (int start_row = 0; start_row < nrow; start_row++){
            int[] cur_row = new int[ncol + 1];
            int[] prev_row = new int[ncol + 1];
            for(int row = start_row; row < nrow;  row++){
                TreeSet<Integer> ts = new TreeSet<>();
                for(int col = 0; col < ncol; col++){
                    cur_row[col + 1] = cur_row[col] + prev_row[col+1] - prev_row[col] +matrix[row][col];
                    ts.add(cur_row[col]);
                   // System.out.println(cur_row[col+1]);
                    int target = cur_row[col + 1] - k;
                    if (ts.ceiling(target) != null){
                        int val = ts.ceiling(target);
                        res = Math.max(cur_row[col + 1]-val, res);
                    }
                    if(res == k) return k;

                }
                prev_row = cur_row.clone();
            }
              
        }
        return res;
    }
}